### PR TITLE
bugfixes, styling, and improvements

### PR DESCRIPTION
**Pull request by diegocr**

### CODE REVIEW

**Comment by infinity0** [comment]
Can you split up the commit into separate commits for bugfixing vs style? See `man git-rebase` SPLITTING COMMITS section for some instructions on that.

**Comment by diegocr** [comment]
Hi Ximin, here's the diff with no styling https://gist.github.com/diegocr/adae0b1d77f3521f2f91

**Comment by infinity0** [comment]
Yes, but can you fix up this pull request to incorporate the split commits, please - e.g. using `git push -f origin master` on your local repo. Contributing to an open source project carries certain responsibilities, including making your commits easier to review for others. The commit description messages should be more specific, too. For example, what bugs have you fixed?

File: lib/megaLogger.js [hunk 1]
```diff
@@ -80,12 +78,26 @@
...
+     * @private
+     * @static
+     */
+    var _intToLevel = {
+        '40': 'ERROR',
+        '30': 'WARN',
+        '20': 'INFO',
+        '10': 'LOG',
+        '0': 'DEBUG'
+    };
```
Comment (diegocr) [inline lib/megaLogger.js:90]: Here i replaced the `_intToLevel()` function by an _object_, which should be much faster than traversing the `LEVELS` in every logging invocation.

File: lib/megaLogger.js [hunk 2]
```diff
@@ -80,12 +78,26 @@
...
+
+    /**
      * Static, global log registry
      *
      * @private
      * @static
      */
-    MegaLogger._logRegistry = {};
+    var _logRegistry = {};
```
Comment (diegocr) [inline lib/megaLogger.js:98]: Since that is `@private`, it doesn't really need to be public. I've made this change in a couple of other places as you'll see below.

File: lib/megaLogger.js [hunk 3]
```diff
@@ -94,27 +106,19 @@
...
-        else if (typeof(window) !== 'undefined' && /PhantomJS/.test(window.navigator.userAgent)) {
-            isBrowser = false;
-        } else if (
-            typeof(window) !== 'undefined' && (
-            /MSIE/.test(window.navigator.userAgent) ||
-            /Trident/.test(window.navigator.userAgent) ||
-            /Edge/.test(window.navigator.userAgent)
-            )
-        ) {
+        else if ("ActiveXObject" in window) {
```
Comment (diegocr) [inline lib/megaLogger.js:114]: Edge/Spartan is based on Webkit, hence it must support colors. MSIE 11 doesn't identify itself with "msie" in the user agent which is why `Trident` was used here i think, however just checking for `ActiveXObject` should be faster/cleaner than dealing with RegEx and strings since it exists in every MSIE version.

File: lib/megaLogger.js [hunk 4]
```diff
@@ -132,23 +136,21 @@
             'LOG': '#000000'
         },
         'dateFormatter': function(d) {
-            return d.getFullYear() + "-" + d.getMonth() + "-" + d.getDate() + " " + d.getHours() + ":" + d.getMinutes() + ":" + d.getSeconds() + "." + d.getMilliseconds();
+            return d.toISOString();
```
Comment (diegocr) [inline lib/megaLogger.js:137]: Here an ISO 8601 date was being mistakenly created, with a wrong month and hour digits with no preceding zeros.

So, just replacing that with `toISOString()` should do the same in a perfect way.

File: lib/megaLogger.js [hunk 5]
```diff
@@ -132,23 +136,21 @@
...
         },
-        'transport': function() {
-            var level = Array.prototype.slice.call(arguments, 0);
-            var args = level.splice(1);
+        'transport': function(level, args) {
```
Comment (diegocr) [inline lib/megaLogger.js:139]: Here i wasn't sure what was the purpose of dealing with Array's `slice` & `splice` functions when you can just provide the plain named arguments :confused: so, this should be a little less overkill.

File: lib/megaLogger.js [hunk 6]
```diff
@@ -252,55 +232,48 @@
...
-            (self.options.colorsEnabled ? "%c" : "") + self.options.dateFormatter(new Date()) + " - " + self._getLoggerPath() + " - " + levelName,
-            (self.options.colorsEnabled ? logStyle : "")
-        ];
-
-        args = args.concat(arguments);
-        if(self.options.dereferenceObjects) {
-            args = JSON.parse(JSON.stringify(args));
+        // check min log level
+        if (level < options.minLogLevel()) {
+            return;
```
Comment (diegocr) [inline lib/megaLogger.js:241]: Here too many stuff was being unnecessarily performed **if** the level isn't high enough or logging is disabled. So, we will be now checking the level first, and i've also moved all that within the `isEnabled` block.

File: lib/megaLogger.js [hunk 7]
```diff
@@ -252,55 +232,48 @@
...
             }
 
-            if(level < self.options.minLogLevel()) { // check min log level
-                return;
+            var muteList = options.muteList();
+            if (muteList.length) {
+                var txtMsg = args.join(" ");
+                if (muteList.some(function(v) { return RegExp(v).test(txtMsg); })) {
+                    return;
+                }
```
Comment (diegocr) [inline lib/megaLogger.js:296]: This is just a tiny improvement for the `muteList`, relying on Array's `some` over `forEach`...

File: lib/megaLogger.js [hunk 8]
```diff
@@ -311,34 +284,34 @@
...
+                    fn = "debug";
+                } else if (level === MegaLogger.LEVELS.LOG) {
+                    fn = "log";
+                } else if (level === MegaLogger.LEVELS.INFO) {
+                    fn = "info";
+                } else if (level === MegaLogger.LEVELS.WARN) {
+                    fn = "warn";
+                } else if (level === MegaLogger.LEVELS.ERROR) {
+                    fn = "error";
                 }
```
Comment (diegocr) [inline lib/megaLogger.js:331]: NB: all this block makes no sense, since this is only reached when `level === MegaLogger.LEVELS.ERROR`

**Comment by infinity0** [comment]
Thanks, the comments are useful and will make my time easier. However, what I meant is that they should be incorporated into the _commit message_ and be split out over several commits, to make it easy for future people to see what's going on as well. The existence of this pull request is not going to be easy to find (and we don't want to depend on GitHub existing in the future), whereas it's easy to run `git log` or `git blame` etc. Even though this project is small, it's good practise to get into this sort of habit for larger projects.

File: lib/megaLogger.js (hunk 1 above)
Comment (lpetrov) [inline lib/megaLogger.js:90]: The main idea here was to have one simple place that you can add/remove additional log levels (e.g. in MegaLogger.LEVELS).
However, i agree that the resources required for this makes this a bad idea :)

TLDR; nice, i will merge this :)

File: lib/megaLogger.js (hunk 2 above)
Comment (lpetrov) [inline lib/megaLogger.js:98]: MegaLogger._logRegistry was initially used as this, so that i can access it via the debugger/web dev tools to debug some loggers.... but now, after the development of this project is mainly into "maintenance mode", i like the change :) :+1:

File: lib/megaLogger.js (hunk 3 above)
Comment (lpetrov) [inline lib/megaLogger.js:114]: I'm not a fan of implicit checks like this one, but ... i will merge it, it makes sense in terms of < RegExp/resources when checking.

Note: I will also add a comment to that line referring to this merge request.

File: lib/megaLogger.js (hunk 4 above)
Comment (lpetrov) [inline lib/megaLogger.js:137]: :+1:

File: lib/megaLogger.js (hunk 5 above)
Comment (lpetrov) [inline lib/megaLogger.js:139]: :+1:

originally, i had the idea of the 'transport' func to get called with some a bit more dynamic arguments, but ... it did not made sense after a while, so nice catch :)

File: lib/megaLogger.js (hunk 7 above)
Comment (lpetrov) [inline lib/megaLogger.js:296]: :+1:

File: lib/megaLogger.js (hunk 8 above)
Comment (lpetrov) [inline lib/megaLogger.js:331]: :+1: 
will fix it in the merge.
